src/main/java/org/testng/TestRunner.java

```diff
@@ -707,7 +716,8 @@ private void privateRun(XmlTest xmlTest) {
                   threadCount,
                   0,
                   TimeUnit.MILLISECONDS,
-                  new LinkedBlockingQueue<>());
+                  new PriorityBlockingQueue<>(),
```

Comment:
This is a big part of the fix. Using a priority queue here means that as we add new methods to run, they'll be sorted by priority with the other things that are waiting to be run. However, that only seems to apply when this queue isn't empty and there aren't worker threads waiting to grab things and run with them.

Reply:
@krmahadevan Agree, IMO we should not change dependencies order

Reply:
I'm trying to not change the dependency order. Let me try to explain a little more:

In single-threaded operation, my change should not affect test run order in any way. That was my goal. The behavior that I'm trying to improve is when you run with `parallel=METHOD`. In the parallel case, the current TestNG behavior is that all methods of a higher priority must **finish** before methods of a lower priority can **start**. When using multiple threads, this can create a substantial inefficiency: if I have 5 threads, 10 test methods total, and 2 slow test methods with higher priority than everything else, those 2 slow methods will be run first, and none of the lower-priority methods will start until both of those slow methods finish.

I think that, instead, the behavior should be that all methods of a higher priority are **started** before methods of a lower priority can **start**. So in my example above, I would expect to see the 2 high-priority methods started first, then 3 of the regular methods would start in order to take full advantage of the 5 threads. Those 3 regular methods have **no dependencies** on the 2 slow methods, so there's no reason they can't be started while the 2 slow methods are running.

Reply:
You made a few more points that I didn't address. Re: https://github.com/cbeust/testng/pull/1910#discussion_r214503899, priority only comes into play when looking at several tests that either don't have dependencies, or their dependencies have been fulfilled. All 4 of the dependencies that you mentioned, soft or hard, will take precedence over priority. A high-priority test method with an unfulfilled dependency will still never run before its dependency has been fulfilled.

And to address https://github.com/cbeust/testng/pull/1910#discussion_r214503546, yes, if there is no priority specified anywhere, then we can just use the `LinkedBlockingQueue`, instead. I'll look into implementing that.